### ISSUE TITLE
Fix 'NoneType' is not iterable TypeError for `salt-run cache.grains`

### ISSUE THREAD

**Issue by myii** [saltstack/salt#39835]
### Description of Issue/Question

Possible duplicate of #38216.  Unfortunately, it seems that this problem persists, even after fix #38659 linked to that issue.

```salt
# salt-run cache.grains
Exception occurred in runner cache.grains: Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/salt/client/mixins.py", line 395, in _low
    data['return'] = self.functions[fun](*args, **kwargs)
  File "/usr/lib/python2.7/dist-packages/salt/runners/cache.py", line 38, in grains
    cached_grains = pillar_util.get_minion_grains()
  File "/usr/lib/python2.7/dist-packages/salt/utils/master.py", line 292, in get_minion_grains
    cached_minion_grains, cached_minion_pillars = self._get_cached_minion_data(*minion_ids)
  File "/usr/lib/python2.7/dist-packages/salt/utils/master.py", line 137, in _get_cached_minion_data
    if 'grains' in mdata:
TypeError: argument of type 'NoneType' is not iterable
```

Checked the main docs @ https://docs.saltstack.com/en/latest/ref/runners/all/salt.runners.cache.html:

```salt
 salt.runners.cache.grains(tgt=None, expr_form='glob', **kwargs)
```

Found that explicitly stating the `tgt` worked fine, i.e.:

```salt
# salt-run cache.grains tgt='*'
```

This gave the result that `salt-run cache.grains` was expected to provide.

### Versions Report
```bash
# salt --versions-report
Salt Version:
           Salt: 2016.11.3
 
Dependency Versions:
           cffi: 0.8.6
       cherrypy: 3.2.2
       dateutil: 1.5
          gitdb: 0.5.4
      gitpython: 0.3.2 RC1
          ioflo: Not Installed
         Jinja2: 2.7.2
        libgit2: Not Installed
        libnacl: Not Installed
       M2Crypto: 0.21.1
           Mako: 0.9.1
   msgpack-pure: Not Installed
 msgpack-python: 0.4.6
   mysql-python: 1.2.3
      pycparser: 2.10
       pycrypto: 2.6.1
         pygit2: Not Installed
         Python: 2.7.6 (default, Oct 26 2016, 20:30:19)
   python-gnupg: Not Installed
         PyYAML: 3.10
          PyZMQ: 14.0.1
           RAET: Not Installed
          smmap: 0.8.2
        timelib: Not Installed
        Tornado: 4.2.1
            ZMQ: 4.0.5
 
System Versions:
           dist: Ubuntu 14.04 trusty
        machine: x86_64
        release: 3.13.0-66-generic
         system: Linux
        version: Ubuntu 14.04 trusty
```

**Comment by myii** [comment]
I have another `2016.11.3` master available, which doesn't have this problem.

```bash
# salt --versions-report
Salt Version:
           Salt: 2016.11.3
 
Dependency Versions:
           cffi: Not Installed
       cherrypy: Not Installed
       dateutil: 1.5
          gitdb: 0.5.4
      gitpython: 0.3.2 RC1
          ioflo: Not Installed
         Jinja2: 2.7.2
        libgit2: Not Installed
        libnacl: Not Installed
       M2Crypto: Not Installed
           Mako: 0.9.1
   msgpack-pure: Not Installed
 msgpack-python: 0.4.6
   mysql-python: 1.2.3
      pycparser: Not Installed
       pycrypto: 2.6.1
         pygit2: Not Installed
         Python: 2.7.6 (default, Oct 26 2016, 20:30:19)
   python-gnupg: Not Installed
         PyYAML: 3.10
          PyZMQ: 14.0.1
           RAET: Not Installed
          smmap: 0.8.2
        timelib: Not Installed
        Tornado: 4.2.1
            ZMQ: 4.0.5
 
System Versions:
           dist: Ubuntu 14.04 trusty
        machine: x86_64
        release: 3.16.0-67-generic
         system: Linux
        version: Ubuntu 14.04 trusty
```

The only differences, in terms of the output of this command, are:

1. `cffi`
1. `cherrypy`
1. `M2Crypto`
1. `pycparser`
1. `release`

Hope that helps narrow things down a bit.

**Comment by Ch3LL** [comment]
@myii can you give this a try https://github.com/saltstack/salt/pull/39852

This is most likely a duplicate of https://github.com/saltstack/salt/issues/39786

**Comment by myii** [comment]
@Ch3LL Thanks for for your reply.  I'll try this out within the next day and reply here.

**Comment by myii** [comment]
@Ch3LL OK, applying that patch has worked, so thanks for that.

There is point of interest, though.  I've checked my second master (as mentioned above, which was already working) and it **doesn't** have this patch applied and is working regardless.  So I've run a comparison of `/usr/lib/python2.7/dist-packages/salt` on both masters and the diff shows that they were identical, before I applied the patch.

Is there anything here to chase up further or shall I close the issue?
